bftengine/src/communication/TlsTCPCommunication.cpp

```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+                                     << ", dest: " << _destId
+                                     << ", connected: " << _connected
+                                     << "is_open: " << get_socket().is_open());
   }
 
-  void read_header_async() {
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId
-                             << ", connected: " << connected << "is_open: "
-                             << get_socket().is_open());
-
-    memset(_inBuffer, 0, _bufferLength);
-    async_read(*_socket,
-               buffer(_inBuffer, LENGTH_FIELD_SIZE),
-               boost::bind(&AsyncTlsConnection::read_header_async_completed,
-                           shared_from_this(),
-                           boost::asio::placeholders::error,
-                           boost::asio::placeholders::bytes_transferred));
-
-    LOG_TRACE("exit, node " << _selfId << ", dest: " << _destId
-                            << ", connected: " << connected << "is_open: "
-                            << get_socket().is_open());
-  }
+  /**
+   * start reading message length bytes from the stream
+   */
+  void read_msg_length_async() {
+    if (_closed)
+      return;
 
-  bool is_service_message() {
-    uint16_t msgType =
-        *(static_cast<uint16_t *>(
-            static_cast<void *>(_inBuffer + LENGTH_FIELD_SIZE)));
-    switch (msgType) {
-      case MessageType::Hello:
-        _destId =
-            *(static_cast<NodeNum *>(
-                static_cast<void *>(
-                    _inBuffer + LENGTH_FIELD_SIZE +
-                        MSGTYPE_FIELD_SIZE)));
-
-        LOG_DEBUG("node: " << _selfId << " got hello from:" << _destId);
-
-        _fOnHellOMessage(_destId, shared_from_this());
-        return true;
-        break;
-      default:return false;
-    }
-  }
+    // since we allow partial reading here, we dont need timeout
+    _socket->async_read_some(
+        asio::buffer(_inBuffer, MSG_LENGTH_FIELD_SIZE),
+        boost::bind(&AsyncTlsConnection::read_msglength_completed,
+                    shared_from_this(),
+                    boost::asio::placeholders::error,
+                    boost::asio::placeholders::bytes_transferred,
+                    true));
+
+    LOG_DEBUG(_logger,
+              "read_msg_length_async, node " << _selfId
+               << ", dest: " << _destId
+               << ", connected: " << _connected
+               << "is_open: " << get_socket().is_open());
+}
 
+  /**
+   * occurs when the whole message has been read from the stream
+   * @param ec error code
+   * @param bytesRead  actual bytes read
+   */
   void read_msg_async_completed(const boost::system::error_code &ec,
                                 size_t bytesRead) {
-    // (IG): patch, dont do it, need to fix multithreading
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId);
-
-    lock_guard<recursive_mutex> lock(_connectionsGuard);
+    auto res = _readTimer.expires_at(boost::posix_time::pos_infin);
```

Comment:
Well in this case both operations have the same effect - because we have only 1 thread and in this point of time there may be at most 1 pending async_wait operation that need to be cancelled. I think semantically it's better to say "here I don't want to cancel the timer, I just want it not to expire because I have all my data been read" and in close() the timer is cancelled totally.